### DISCUSSION TITLE
Integrating Custom JWTs with Supabase RLS in Next.js Middleware

### DISCUSSION THREAD

**Discussion by codinesh** [supabase/supabase#21705] [Questions] [answered]
**Description:**

Facing challenges setting up RLS based on custom JWTs from a 3rd-party OAuth2 provider in a Next.js application.
Tried setting Supabase auth session and global headers, but encountering issues.

I am using a 3rd party oauth2 provider, which generates JWT token, I want to set RLS policy against this token and set the token in nextjs middleware.

1. Auth Session Not Working:
supabase.auth.setSession(...) doesn't include the token in subsequent calls.
   ```ts
      // middleware.ts
      supabase.auth.setSession({
            access_token: customToken,
            refresh_token: "",
          });
    ```
        
2. 409 Conflict with Global Header:
Setting Authorization header in createServerClient options results in 409 errors.
Edge logs confirm user_id from provided token.
    ```ts
           if (token) {
          options.auth = {
            autoRefreshToken: false,
          };
          options.global = {
            headers: {
              Authorization: `Bearer ${token}`,
            },
          };
        }
      
        const supabase =  createServerClient(
          process.env.NEXT_PUBLIC_SUPABASE_URL!,
          process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
          options
        );
    ```
    
    
    #21160 seems to be similar but was using realtime. 
    
    Please suggest if this is correct approach or any pointers.

**Comment by GaryAustin1** [comment]
Probably need to spell a bit more on what you are doing with custom jwt's.

In general you can't use auth-helpers/SSR with custom JWTs.  This method is designed to work with Supabase Auth and sessions including the refresh tokens.

If you do custom JWT's then you can use a basic supabase-js client and the Authorization header with a JWT signed by the JWT secret.  But you can't mix that with auth-helpers methods.

**Reply by codinesh** [reply]
Thanks @GaryAustin1 for or the response. I will try with the basic js sdk.

Here is the details on what I am doing with jwt token.

Token received from third party auth provider has certain claims as mentioned in example below, with "alg": "ES256". 

I am generating another token with the same claims with "alg": "HS256" and with JWT secret from supabase. Passing the new token to supabase. Seems like the userid from this token is validated against the users table in auth schema (which don't have any records, since I am not using supabase auth) and returning 409.

```json
 {
  appId: "<apid>",
  issuer: "privy.io",
  issuedAt: 1709354701,
  expiration: 1709358301,
  sessionId: "<session>",
  userId: "<userid>",
}
```

RLS policy for Insert on a table. (Hardcoded with a test user id to check the integration, later will use a role or other check)

```json
BEGIN;
  ALTER POLICY "Enable insert for admin users only" ON "public"."profile" WITH CHECK (((auth.jwt() ->> 'sub'::text) = '<testuserid>'));
COMMIT;```

**Reply by GaryAustin1** [reply, marked as answer]
So the users table is not involved at all if you don't use Supabase Auth.  All that matters are the JWT claims and what you test in RLS of them (if the JWT is signed and not expired).  Normally doing custom JWT claims you would use auth.jwt()->>'claimName' in the JWT for claims.  Role claim is required to be anon/authenticated/service_role without custom Postgres roles.  Sub is normally used by auth.uid() but it must be a UUID to use that function.  You can also do auth.jwt()->>'sub' on a text claim if that is what your Auth provider uses.

**Reply by codinesh** [reply]
okay, thanks for clarification. Do you have any thought on the 409 response from supabase call.

From the API Edge Network logs in supabase portal, I can see that the headers has the auth_user property which is set to value of sub in token. Does this mean the token parsing is working and RLS check fails? Since there is no other supabse tables involved, I am confused on the 409 conflict response. 

```json
"sb": [
    {
      "auth_user": "<userid passed from token>"
    }
  ],
```

**Reply by GaryAustin1** [reply]
It is late for me.  But I don't think you have posted your call that gets 409.  

409 does not seem to me involving RLS.

**Reply by codinesh** [reply]
> It is late for me.  But I don't think you have posted your call that gets 409.  
> 
> 
> 
> 409 does not seem to me involving RLS.  
> 
> 
> 
> 

Please take rest. I will look for solution based on your suggestions. Thanks for your time.

**Reply by codinesh** [reply]
This is working, conflict was not due to RLS as pointed by @GaryAustin1, but due to constraints on table.
